src/main/java/fi/metatavu/metamind/server/rest/StoriesApiImpl.java

```diff
@@ -456,40 +608,101 @@ public Response findVariable(UUID storyId, UUID variableId) {
     if (story == null) {
       return createBadRequest(String.format("Story %s not found", storyId)); 
     }
     
+    List<fi.metatavu.metamind.persistence.models.Intent> intents = storyController.listIntentsByStory(story);
+    if (intents.isEmpty()) {
+      return createNotFound(String.format("Intents %s not found", storyId)); 
+    }
+    
+    try {
+      for (fi.metatavu.metamind.persistence.models.Intent intent : intents) {
+        UUID foundResourceId = authenticationController.findProtectedResource(String.format(INTENT_NAME_BEGINNING, intent.getId()));
+        Set<UUID> permittedUsersId = authenticationController.getResourcePermittedUsers(REALM_NAME, foundResourceId, String.format(INTENT_NAME_BEGINNING, intent.getId()), SCOPES);
+        if (!permittedUsersId.contains(loggedUserId) || loggedUserId == null) {
+          return createBadRequest(String.format(NO_USER_FOUND_MESSAGE));
+        }
+      }
+      
+    } catch (Exception e) {
+      logger.warn(String.format(EXCEPTION_CAUGHT, e.getMessage()));
+    }
+    
     return createOk(storyController.listIntentsByStory(story).stream()
       .map(intent -> intentTranslator.translateIntent(intent, trainingMaterialController.listTrainingMaterialByIntent(intent)))
       .collect(Collectors.toList()));
   }
 
   @Override
   public Response listKnots(UUID storyId) {
-    // TODO: Permission check
     
     fi.metatavu.metamind.persistence.models.Story story = storyController.findStoryById(storyId);
     if (story == null) {
       return createBadRequest(String.format("Story %s not found", storyId)); 
     }
     
+    List<fi.metatavu.metamind.persistence.models.Knot> knots = storyController.listKnotsByStory(story);
+    if (knots.isEmpty()) {
+      return createNotFound(String.format("Knots %s not found", storyId)); 
+    }
+    
+    try {
+      for (fi.metatavu.metamind.persistence.models.Knot knot : knots) {
+        UUID foundResourceId = authenticationController.findProtectedResource(String.format(KNOT_NAME_BEGINNING, knot.getId()));
+        Set<UUID> permittedUsersId = authenticationController.getResourcePermittedUsers(REALM_NAME, foundResourceId, String.format(KNOT_NAME_BEGINNING, knot.getId()), SCOPES);
+        if (!permittedUsersId.contains(loggedUserId) || loggedUserId == null) {
+          return createBadRequest(String.format(NO_USER_FOUND_MESSAGE));
+        }
+      }
+      
+    } catch (Exception e) {
+      logger.warn(String.format(EXCEPTION_CAUGHT, e.getMessage()));
+    }
+    
     return createOk(storyController.listKnotsByStory(story).stream()
       .map(knotTranslator::translateKnot)
       .collect(Collectors.toList()));
   }
 
   @Override
   public Response listStories() {
-    // TODO: Permission check
+    
+    List<fi.metatavu.metamind.persistence.models.Story> stories = storyController.listStories();
+    if (stories == null) {
+      return createBadRequest("Stories %s not found"); 
+    }
+    
+    try {
```

Comment:
This could create severe permission issues when the realm has thousands of users, the list should be filtered with evaluation requests into Keycloak instead of resolving all permitted users and comparing it against all users in the realm